### PR TITLE
Validation

### PR DESCRIPTION
**Pull request by dialelo**
Opening a PR for requesting comments, still a WIP. I'm not sure whether this belongs in cats, maybe the type is OK but I'm thinking on implementing a schema-like library with lots of combinators for validating data structures; I'm also not sure if creating a new `cats.applicative` namespace is a good idea.

Still to-do:
- [x] Semigroup & Monoid instances of Validation
- [x] More thorough testing
- [x] Documentation

### CODE REVIEW

**Comment by niwibe** [comment]
Maybe the proper approach is implement everything related to abstraction in cats but the "schema-like" library as separate one. Would be awesome have one, very great idea!

**Comment by dialelo** [comment]
This is it, two questions remain.
- Are you OK with the `cats.applicative` namespace?
- Is `vale` a good name for the aforementioned validation library? (alternatives: valorio, valiente)

**Comment by niwibe** [comment]
I think that applicative ns is appropriate, because types defined in that namespace are applicative and not monad...

And the `vale` name convinces me more that the alternatives :P

**Comment by dialelo** [comment]
Awesome, it'd be great if you push a snapshot to Clojars after merging this PR, I'm going to start sketching the `vale` library :shipit:

**Comment by niwibe** [comment]
It is still WIP or is already finished?

I have reviewed it and it seems ok for me. Feel free to merge it yourself ;)

**Comment by niwibe** [comment]
funcool/cats "0.5.0-SNAPSHOT"

**Comment by dialelo** [comment]
:cake: thanks!

File: doc/cats.adoc [hunk 1]
```diff
@@ -675,18 +675,67 @@ TODO
...
+It implements the Applicative protocol, and its intended usage is as an Applicative. Applying Validation
+values together errs on the side of the failure, and applying failures together aggregates their values
+using the Semigroup's `mappend` function.
+
+[source, clojure]
+----
+(require '[cats.applicative.validation :as v])
+(require '[cats.core :as m])
+
+(m/fapply (v/ok 42) (v/fail "OH NOES"))
```
Comment (clyfe) [inline doc/cats.adoc]: This makes no sense. The first arg was supposed to wrap a fn.

Comment (dialelo) [inline doc/cats.adoc]: It does make sense. Note that in the Validation applicative function application is only run with two Ok instances. It [follows the applicative laws](https://github.com/funcool/cats/blob/master/test/cats/applicative/validation_spec.cljc), see [this Haskell library](https://github.com/NICTA/validation) for different flavours of validation types.

**Comment by clyfe** [comment]
```clojure
(m/fapply (v/ok 1) (v/ok 2))
; j.l.ClassCastException: j.l.Long cannot be cast to c.l.IFn
(m/fapply (v/ok inc) (v/ok 2))
; #<Ok 3>
```
Maybe I'm missing something.

File: doc/cats.adoc (hunk 1 above)
Comment (clyfe) [inline doc/cats.adoc]:
```clojure
(m/fapply (v/ok 1) (v/ok 2))
; j.l.ClassCastException: j.l.Long cannot be cast to c.l.IFn
(m/fapply (v/ok inc) (v/ok 2))
; #<Ok 3>
```
Maybe I'm missing something.

**Pull request merged**